src/core/highlighted_text.js

```diff
@@ -0,0 +1,203 @@
+ * limitations under the License.
+ */
+
+class SingleIntersector {
+  #annotation;
+
+  #minX = Infinity;
+
+  #minY = Infinity;
+
+  #maxX = -Infinity;
+
+  #maxY = -Infinity;
+
+  #quadPoints;
+
+  #text = [];
+
+  #extraChars = [];
+
+  #lastIntersectingQuadIndex = -1;
+
+  #canTakeExtraChars = false;
+
+  constructor(annotation) {
+    this.#annotation = annotation;
+    const quadPoints = (this.#quadPoints = annotation.data.quadPoints);
+
+    for (let i = 0, ii = quadPoints.length; i < ii; i += 8) {
+      this.#minX = Math.min(this.#minX, quadPoints[i]);
+      this.#maxX = Math.max(this.#maxX, quadPoints[i + 2]);
+      this.#minY = Math.min(this.#minY, quadPoints[i + 5]);
+      this.#maxY = Math.max(this.#maxY, quadPoints[i + 1]);
+    }
+  }
+
+  overlaps(other) {
+    return !(
+      this.#minX >= other.#maxX ||
+      this.#maxX <= other.#minX ||
+      this.#minY >= other.#maxY ||
+      this.#maxY <= other.#minY
+    );
+  }
+
+  /**
+   * Check if the given point intersects with the annotation's quad points.
+   * The point (x, y) is supposed to be the center of the glyph.
+   * @param {number} x
+   * @param {number} y
+   * @returns {boolean}
+   */
+  #intersects(x, y) {
+    if (
+      this.#minX >= x ||
+      this.#maxX <= x ||
+      this.#minY >= y ||
+      this.#maxY <= y
+    ) {
+      return false;
+    }
+
+    const quadPoints = this.#quadPoints;
+    if (quadPoints.length === 8) {
+      // We've only one quad, so if we intersect min/max bounds then we
+      // intersect the quad.
+      return true;
+    }
+
+    if (this.#lastIntersectingQuadIndex >= 0) {
+      const i = this.#lastIntersectingQuadIndex;
+      if (
+        !(
+          quadPoints[i] >= x ||
+          quadPoints[i + 2] <= x ||
+          quadPoints[i + 5] >= y ||
+          quadPoints[i + 1] <= y
+        )
+      ) {
+        return true;
+      }
+      this.#lastIntersectingQuadIndex = -1;
+    }
+
+    for (let i = 0, ii = quadPoints.length; i < ii; i += 8) {
+      if (
+        !(
+          quadPoints[i] >= x ||
+          quadPoints[i + 2] <= x ||
+          quadPoints[i + 5] >= y ||
+          quadPoints[i + 1] <= y
+        )
+      ) {
+        this.#lastIntersectingQuadIndex = i;
+        return true;
+      }
+    }
+    return false;
+  }
+
+  addGlyph(x, y, glyph) {
+    if (!this.#intersects(x, y)) {
+      this.disableExtraChars();
+      return false;
+    }
+
+    if (this.#extraChars.length > 0) {
+      this.#text.push(this.#extraChars.join(""));
+      this.#extraChars.length = 0;
+    }
+    this.#text.push(glyph);
+    this.#canTakeExtraChars = true;
+
+    return true;
+  }
+
+  addExtraChar(char) {
+    if (this.#canTakeExtraChars) {
+      this.#extraChars.push(char);
+    }
+  }
+
+  disableExtraChars() {
+    if (!this.#canTakeExtraChars) {
+      return;
+    }
+    this.#canTakeExtraChars = false;
+    this.#extraChars.length = 0;
+  }
+
+  getText() {
```

Comment:
With `get` in the name I would have expected this method to return the text instead of setting it on the annotation object. Maybe call it `setText` instead?